src/deadline/job_attachments/api/attachment.py

```diff
@@ -28,56 +33,156 @@ def attachment_download(
+        upload_manifest_path (Optional[str], optional): Optional prefix for uploading given manifests. Defaults to None.
+        logger (ClickLogger, optional): Logger to provide visibility. Defaults to ClickLogger(False).
+
+    Raises:
+        NonValidInputError: _description_
+    """
+
+    if not all(os.path.isdir(root) for root in root_dirs):
+        raise NonValidInputError(f"Specified root dir {root_dirs} contains nonvalid directory.")
+
+    file_name_manifest_dict: Dict[str, BaseAssetManifest] = _get_file_name_manifest_dict(
+        manifests=manifests
+    )
+    path_mapping_rule_list: List[PathMappingRule] = _process_path_mapping(
+        path_mapping_rules=path_mapping_rules, root_dirs=root_dirs
+    )
+
+    s3_settings: JobAttachmentS3Settings = JobAttachmentS3Settings.from_s3_root_uri(s3_root_uri)
+    asset_uploader: S3AssetUploader = S3AssetUploader(session=boto3_session)
+    for file_name in file_name_manifest_dict:
+        manifest: BaseAssetManifest = file_name_manifest_dict[file_name]
+
+        # File name is supposed to be prefixed by a hash of source path in path mapping or provided root dirs
+        rule: Optional[PathMappingRule] = next(
+            # search in path mapping to determine source and destination
+            (
+                rule
+                for rule in path_mapping_rule_list
+                if rule.get_hashed_source_path(manifest.get_default_hash_alg()) in file_name
+            ),
+            None,
+        )
+        if not rule:
+            raise NonValidInputError(
+                f"No valid root defined for given manifest {file_name}, please check input root dirs and path mapping rule."
+            )
+
+        # Uploads all files to a CAS in the manifest, optionally upload manifest file
+        key, data = asset_uploader.upload_assets(
+            job_attachment_settings=s3_settings,
+            manifest=manifest,
+            partial_manifest_prefix=upload_manifest_path,
+            source_root=Path(rule.source_path),
+            asset_root=Path(rule.destination_path),
+        )
+        logger.echo(
+            f"Uploaded assets from {rule.source_path}, to {s3_settings.to_s3_root_uri()} key {key}, hashed data {data}"
+        )
 
-    parsed_mappings: List[PathMappingRule] = list()
+
+def _process_path_mapping(
+    path_mapping_rules: Optional[str] = None, root_dirs: List[str] = []
+) -> List[PathMappingRule]:
+
+    path_mapping_rule_list: List[PathMappingRule] = list()
 
     if path_mapping_rules:
         if not os.path.isfile(path_mapping_rules):
             raise NonValidInputError(
                 f"Specified path mapping file {path_mapping_rules} is not valid."
             )
-
         with open(path_mapping_rules, encoding="utf8") as f:
-            parsed_mappings = [PathMappingRule(**mapping) for mapping in json.load(f)]
+            path_mapping_rule_list = [PathMappingRule(**mapping) for mapping in json.load(f)]
```

Comment:
```suggestion
            path_mapping_rule_list.extend([PathMappingRule(**mapping) for mapping in json.load(f)])
```

Since we are initializing the list above now, we should extend here instead of replacing it with a new list